Can't run Protractor e2e on newly generate Angular app

<!-- Please do your best to fill out all of the sections below! -->

## Current Behavior
<!-- What is the behavior that currently you experience? -->

Protractor e2e tests fail on a newly generated Nrwl Angular application with the error:

```
** Angular Live Development Server is listening on localhost:4200, open your browser on http://localhost:4200/ **
: Compiled successfully.
[08:01:26] I/launcher - Running 1 instances of WebDriver
[08:01:26] I/direct - Using ChromeDriver directly...
[08:01:27] E/launcher - Error: TSError: ⨯ Unable to compile TypeScript:
error TS5083: Cannot read file '/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/tsconfig.json'.

    at createTSError (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/ts-node/src/index.ts:261:12)
    at Object.register (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/ts-node/src/index.ts:265:42)
    at onPrepare (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/apps/my-app-e2e/protractor.conf.js:25:24)
    at /Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/built/util.js:46:49
    at Function.promise (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:682:9)
    at Object.runFilenameOrFn_ (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/built/util.js:38:16)
    at /Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/built/runner.js:98:27
    at _fulfilled (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:834:54)
    at /Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:863:30
    at Promise.promise.promiseDispatch (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:796:13)
```

It appears that the e2e tsconfig is extending the wrong file. It should be `"extends": "../../tsconfig.base.json"`. However, after changing this, I receive a different error:

```
** Angular Live Development Server is listening on localhost:4200, open your browser on http://localhost:4200/ **
: Compiled successfully.
[08:08:00] I/launcher - Running 1 instances of WebDriver
[08:08:00] I/direct - Using ChromeDriver directly...
[08:08:02] E/launcher - Error: TypeError: Cannot read property 'PRETTY' of undefined
    at onPrepare (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/apps/my-app-e2e/protractor.conf.js:31:47)
    at /Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/built/util.js:46:49
    at Function.promise (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:682:9)
    at Object.runFilenameOrFn_ (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/built/util.js:38:16)
    at /Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/built/runner.js:98:27
    at _fulfilled (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:834:54)
    at /Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:863:30
    at Promise.promise.promiseDispatch (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:796:13)
    at /Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:556:49
    at runSingle (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:137:13)
```

## Expected Behavior
<!-- What is the behavior that you expect to happen? -->
<!-- Is this a regression? .i.e Did this used to be the behavior at one point?  -->

Protractor e2e tests on a newly generated Nrwl Angular application should succeed.

## Steps to Reproduce
<!-- Help us help you by making it easy for us to reproduce your issue! -->

```
npx create-nx-workspace my-org
cd my-org
yarn add -D -E @nrwl/angular
nx g @nrwl/angular:init
# Make sure to choose Protractor as the e2e test runner
nx g @nrwl/angular:app my-app
nx e2e my-app-e2e
```

<!-- Can you reproduce this on https://github.com/nrwl/nx-examples? -->
<!-- If so, open a PR with your changes and link it below. -->
<!-- If not, please provide a minimal Github repo -->
<!-- At the very least, provide as much detail as possible to help us reproduce the issue -->


### Failure Logs
<!-- Please include any relevant log snippets or files here. -->

```
** Angular Live Development Server is listening on localhost:4200, open your browser on http://localhost:4200/ **
: Compiled successfully.
[08:01:26] I/launcher - Running 1 instances of WebDriver
[08:01:26] I/direct - Using ChromeDriver directly...
[08:01:27] E/launcher - Error: TSError: ⨯ Unable to compile TypeScript:
error TS5083: Cannot read file '/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/tsconfig.json'.

    at createTSError (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/ts-node/src/index.ts:261:12)
    at Object.register (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/ts-node/src/index.ts:265:42)
    at onPrepare (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/apps/my-app-e2e/protractor.conf.js:25:24)
    at /Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/built/util.js:46:49
    at Function.promise (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:682:9)
    at Object.runFilenameOrFn_ (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/built/util.js:38:16)
    at /Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/built/runner.js:98:27
    at _fulfilled (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:834:54)
    at /Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:863:30
    at Promise.promise.promiseDispatch (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:796:13)
```

```
** Angular Live Development Server is listening on localhost:4200, open your browser on http://localhost:4200/ **
: Compiled successfully.
[08:08:00] I/launcher - Running 1 instances of WebDriver
[08:08:00] I/direct - Using ChromeDriver directly...
[08:08:02] E/launcher - Error: TypeError: Cannot read property 'PRETTY' of undefined
    at onPrepare (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/apps/my-app-e2e/protractor.conf.js:31:47)
    at /Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/built/util.js:46:49
    at Function.promise (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:682:9)
    at Object.runFilenameOrFn_ (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/built/util.js:38:16)
    at /Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/built/runner.js:98:27
    at _fulfilled (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:834:54)
    at /Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:863:30
    at Promise.promise.promiseDispatch (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:796:13)
    at /Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:556:49
    at runSingle (/Users/devinshoemaker/code/nx/playground/nx-angular-protractor/node_modules/protractor/node_modules/q/q.js:137:13)
```

### Environment
<!-- It's important for us to know the context in which you experience this behavior! -->
<!-- Please paste the result of `nx report` below! -->

```
>  NX  Report complete - copy this into the issue template

  nx : Not Found
  @nrwl/angular : 10.3.2
  @nrwl/cli : 10.3.2
  @nrwl/cypress : 10.3.2
  @nrwl/eslint-plugin-nx : Not Found
  @nrwl/express : Not Found
  @nrwl/jest : 10.3.2
  @nrwl/linter : 10.3.2
  @nrwl/nest : Not Found
  @nrwl/next : Not Found
  @nrwl/node : Not Found
  @nrwl/react : Not Found
  @nrwl/schematics : Not Found
  @nrwl/tao : 10.3.2
  @nrwl/web : Not Found
  @nrwl/workspace : 10.3.2
  typescript : 4.0.5
```

--- a/packages/angular/migrations.json
+++ b/packages/angular/migrations.json
@@ -40,6 +40,11 @@
       "version": "10.3.0-beta.1",
       "description": "Add tsconfig.editor.json to angular apps and update jest-angular-preset",
       "factory": "./src/migrations/update-10-3-0/update-10-3-0"
+    },
+    "update-10-4-0": {
+      "version": "10.4.0-beta.3",
+      "description": "Adjust karma and protractor setup",
+      "factory": "./src/migrations/update-10-4-0/update-10-4-0"
     }
   },
   "packageJsonUpdates": {
@@ -134,6 +139,47 @@
         },
         "jest-preset-angular": {
           "version": "8.3.1",
+          "alwaysAddToPackageJson": false
+        }
+      }
+    },
+    "10.4.0": {
+      "version": "10.4.0-beta.3",
+      "packages": {
+        "protractor": {
+          "version": "~7.0.0",
+          "alwaysAddToPackageJson": false
+        },
+        "jasmine-core": {
+          "version": "~3.6.0",
+          "alwaysAddToPackageJson": false
+        },
+        "jasmine-spec-reporter": {
+          "version": "~5.0.0",
+          "alwaysAddToPackageJson": false
+        },
+        "@types/jasmine": {
+          "version": "~3.5.0",
+          "alwaysAddToPackageJson": false
+        },
+        "karma": {
+          "version": "~5.0.0",
+          "alwaysAddToPackageJson": false
+        },
+        "karma-chrome-launcher": {
+          "version": "~3.1.0",
+          "alwaysAddToPackageJson": false
+        },
+        "karma-coverage-istanbul-reporter": {
+          "version": "~3.0.2",
+          "alwaysAddToPackageJson": false
+        },
+        "karma-jasmine": {
+          "version": "~4.0.0",
+          "alwaysAddToPackageJson": false
+        },
+        "karma-jasmine-html-reporter": {
+          "version": "^1.5.0",
           "alwaysAddToPackageJson": false
         }
       }

--- a/packages/angular/src/migrations/update-10-4-0/update-10-4-0.spec.ts
+++ b/packages/angular/src/migrations/update-10-4-0/update-10-4-0.spec.ts
@@ -1,0 +1,111 @@
+import { callRule, runMigration } from '../../utils/testing';
+import { chain, Tree } from '@angular-devkit/schematics';
+import {
+  readJsonInTree,
+  updateJsonInTree,
+  updateWorkspace,
+} from '@nrwl/workspace';
+import { createEmptyWorkspace } from '@nrwl/workspace/testing';
+
+describe('update-10.4.0', () => {
+  describe('update Protractor e2e tsconfig.json', () => {
+    let tree: Tree;
+    beforeAll(async () => {
+      tree = Tree.empty();
+      tree = createEmptyWorkspace(tree);
+      tree = await callRule(
+        updateWorkspace((workspace) => {
+          workspace.projects.add({
+            name: 'app1',
+            root: 'apps/app1-e2e',
+            projectType: 'application',
+            targets: {
+              e2e: {
+                builder: '@angular-devkit/build-angular:protractor',
+                options: {
+                  protractorConfig: 'apps/app1-e2e/protractor.conf.js',
+                  devServerTarget: 'testapp:serve',
+                },
+                configurations: {
+                  production: {
+                    devServerTarget: 'testapp:serve:production',
+                  },
+                },
+              },
+            },
+          });
+        }),
+        tree
+      );
+      tree = await callRule(
+        chain([
+          updateJsonInTree('apps/app1-e2e/tsconfig.json', () => ({
+            extends: '../../tsconfig.json',
+          })),
+        ]),
+        tree
+      );
+
+      tree = await runMigration('update-10-4-0', {}, tree);
+    });
+
+    it('should update the protractor e2e tsconfig to correctly point to the tsconfig.base.json', async () => {
+      const tsconfig = readJsonInTree(tree, 'apps/app1-e2e/tsconfig.json');
+      expect(tsconfig.extends).toEqual('../../tsconfig.base.json');
+    });
+  });
+
+  describe('update Protractor, Karma and Jasmine dependencies', () => {
+    let tree: Tree;
+    beforeAll(async () => {
+      tree = Tree.empty();
+      tree = createEmptyWorkspace(tree);
+
+      tree = await callRule(
+        chain([
+          // update package.json to have the deps that are being upgraded
+          updateJsonInTree('package.json', (json) => {
+            json.devDependencies = {
+              ...json.devDependencies,
+              protractor: '1.0.0',
+              'jasmine-core': '1.0.0',
+              'jasmine-spec-reporter': '1.0.0',
+              '@types/jasmine': '1.0.0',
+              karma: '1.0.0',
+              'karma-chrome-launcher': '1.0.0',
+              'karma-coverage-istanbul-reporter': '1.0.0',
+              'karma-jasmine': '1.0.0',
+              'karma-jasmine-html-reporter': '1.0.0',
+            };
+            return json;
+          }),
+        ]),
+        tree
+      );
+
+      tree = await runMigration('update-10-4-0', {}, tree);
+    });
+
+    it('should update the karma, jasmine and protractor dependencies', async () => {
+      const packageJson = readJsonInTree(tree, 'package.json');
+
+      expect(packageJson.devDependencies['protractor']).toEqual('~7.0.0');
+      expect(packageJson.devDependencies['jasmine-core']).toEqual('~3.6.0');
+      expect(packageJson.devDependencies['jasmine-spec-reporter']).toEqual(
+        '~5.0.0'
+      );
+      expect(packageJson.devDependencies['@types/jasmine']).toEqual('~3.5.0');
+      expect(packageJson.devDependencies['karma']).toEqual('~5.0.0');
+      expect(packageJson.devDependencies['karma-chrome-launcher']).toEqual(
+        '~3.1.0'
+      );
+      expect(
+        packageJson.devDependencies['karma-coverage-istanbul-reporter']
+      ).toEqual('~3.0.2');
+      expect(packageJson.devDependencies['karma-jasmine']).toEqual('~4.0.0');
+      expect(
+        packageJson.devDependencies['karma-jasmine-html-reporter']
+      ).toEqual('^1.5.0');
+    });
+  });
+});

--- a/packages/angular/src/migrations/update-10-4-0/update-10-4-0.ts
+++ b/packages/angular/src/migrations/update-10-4-0/update-10-4-0.ts
@@ -1,0 +1,47 @@
+import { normalize, Path } from '@angular-devkit/core';
+import { ProjectDefinition } from '@angular-devkit/core/src/workspace';
+import { chain, Rule, Tree } from '@angular-devkit/schematics';
+import {
+  formatFiles,
+  getWorkspace,
+  offsetFromRoot,
+  updateJsonInTree,
+  updatePackagesInPackageJson,
+} from '@nrwl/workspace';
+import { join as pathJoin } from 'path';
+
+function updateBaseConfig(project: ProjectDefinition, baseConfig: Path): Rule {
+  return updateJsonInTree(baseConfig, (json) => {
+    return {
+      ...json,
+      extends: `${offsetFromRoot(project.root)}tsconfig.base.json`,
+    };
+  });
+}
+
+async function udpateProtractorTsConfig(host: Tree) {
+  const workspace = await getWorkspace(host);
+  const rules = [];
+
+  workspace.projects.forEach((project) => {
+    project.targets.forEach((target) => {
+      if (target.builder === '@angular-devkit/build-angular:protractor') {
+        rules.push(
+          updateBaseConfig(project, normalize(`${project.root}/tsconfig.json`))
+        );
+      }
+    });
+  });
+  return chain(rules);
+}
+
+export default function () {
+  return chain([
+    updatePackagesInPackageJson(
+      pathJoin(__dirname, '../../../migrations.json'),
+      '10.4.0'
+    ),
+    udpateProtractorTsConfig,
+    formatFiles(),
+  ]);
+}

--- a/packages/angular/src/schematics/application/application.ts
+++ b/packages/angular/src/schematics/application/application.ts
@@ -632,6 +632,14 @@ function updateE2eProject(options: NormalizedSchema): Rule {
           };
         }
       ),
+      updateJsonInTree(`${options.e2eProjectRoot}/tsconfig.json`, (json) => {
+        return {
+          ...json,
+          extends: `${offsetFromRoot(
+            options.e2eProjectRoot
+          )}tsconfig.base.json`,
+        };
+      }),
     ]);
   };
 }

--- a/packages/angular/src/schematics/init/init.ts
+++ b/packages/angular/src/schematics/init/init.ts
@@ -91,10 +91,10 @@ export function addE2eTestRunner(options: Pick<Schema, 'e2eTestRunner'>): Rule {
         return addDepsToPackageJson(
           {},
           {
-            protractor: '~5.4.0',
-            'jasmine-core': '~2.99.1',
-            'jasmine-spec-reporter': '~4.2.1',
-            '@types/jasmine': '~2.8.6',
+            protractor: '~7.0.0',
+            'jasmine-core': '~3.6.0',
+            'jasmine-spec-reporter': '~5.0.0',
+            '@types/jasmine': '~3.5.0',
             '@types/jasminewd2': '~2.0.3',
           }
         );

--- a/packages/angular/src/schematics/karma/karma.ts
+++ b/packages/angular/src/schematics/karma/karma.ts
@@ -12,14 +12,14 @@ export default function () {
       addDepsToPackageJson(
         {},
         {
-          karma: '~4.0.0',
-          'karma-chrome-launcher': '~2.2.0',
-          'karma-coverage-istanbul-reporter': '~2.0.1',
-          'karma-jasmine': '~1.1.2',
-          'karma-jasmine-html-reporter': '^0.2.2',
-          'jasmine-core': '~2.99.1',
-          'jasmine-spec-reporter': '~4.2.1',
-          '@types/jasmine': '~2.8.8',
+          karma: '~5.0.0',
+          'karma-chrome-launcher': '~3.1.0',
+          'karma-coverage-istanbul-reporter': '~3.0.2',
+          'karma-jasmine': '~4.0.0',
+          'karma-jasmine-html-reporter': '^1.5.0',
+          'jasmine-core': '~3.6.0',
+          'jasmine-spec-reporter': '~5.0.0',
+          '@types/jasmine': '~3.5.0',
         }
       ),
     ]);

